sips/reduce_signatures_no_fork.md

```diff
@@ -0,0 +1,129 @@
+	}
+
+	// Removed
+	/*
+	if err := signedMsg.GetSignature().VerifyByOperators(signedMsg, b.Share.DomainType, types.PartialSignatureType, b.Share.Committee); err != nil {
+		return errors.Wrap(err, "failed to verify PartialSignature")
+	}
+
+	for _, msg := range signedMsg.Message.Messages {
+		if err := b.verifyBeaconPartialSignature(msg); err != nil {
+			return errors.Wrap(err, "could not verify Beacon partial Signature")
+		}
+	}
+	*/
+	return nil
+}
+```
+
+In the duty runners' `ProcessPostConsensus` and `ProcessPreConsensus` functions, once a quorum is reached, we should attempt to reconstruct and verify the validator's signature. If an error is received, we must fall back, as shown below.
+
+```go
+// Example of post-consensus function for the aggregator duty runner
+func (r *AggregatorRunner) ProcessPostConsensus(signedMsg *types.SignedPartialSignatureMessage) error {
+
+	// ...
+
+	if !quorum {
+		return nil
+	}
+
+	for _, root := range roots {
+		sig, err := r.GetState().ReconstructBeaconSig(r.GetState().PostConsensusContainer, root, r.GetShare().ValidatorPubKey)
+		if err != nil {
+			// Removed
+			/*
+			return errors.Wrap(err, "could not reconstruct selection proof sig")
+			*/
+
+			// New
+			// If reconstructing and verification failed, fall back to verifying each partial signature
+			r.BaseRunner.FallBackToVerifyingEachSignature(r.GetState().PostConsensusContainer, root)
+			return errors.Wrap(err, "got post-consensus quorum but it has invalid signatures")
+		}
+		// ...
+	}
+	// ...
+}
+
+// Fall back function
+func (b *BaseRunner) FallBackToVerifyingEachSignature(container PartialSigContainer, root [32]byte) {
+
+	signatures := container.GetSignatures(root)
+
+	for operatorID, signature := range signatures {
+		if err := b.verifyBeaconPartialSignature(operatorID, signature, root); err != nil {
+			container.Remove(root, operatorID)
+		}
+	}
+}
+```
+
+### Consensus messages
+
+The modification required in the consensus' *prepare* and *commit* functions is equivalent to the one shown in the [partial signature messages processing section](#partial-signature-messages-processing).
+
+## Drawbacks
+
+In the current version, we verify the BLS signature of each message once it is received and, thus, we also drop it immediately if it's wrong. So, if we were to receive two signatures from the same signer, we would store only the correct one (if any). With this proposal, we would need to decide what to do with the signatures, since we don't know which is correct. For that, we could follow one of the following approaches:
+1. verify the duplicated signatures and store the single correct one (if any).
+2. randomly drop one of them.
+3. store and try both until a quorum of signers is reached.
+
+The first approach implies signature verification (what we are trying to reduce), but it's the safest one and, thus, it should be the one adopted. Also, this represents an uncommon case and, therefore, the practical effects are negligible.
```

Comment:
you can write a word that msg validation should prevent this, so it is best to raise an error

Reply:
Msg validation already prevents a node from sending two diff partial signatures for the same duty step from the same signer. Though different nodes could sent two diff partial signatures for the same duty step from the same signer. But I think there's no way of stopping that without verifying the signatures.